sapp/spritebatch-sapp.c

```diff
@@ -0,0 +1,399 @@
+//------------------------------------------------------------------------------
+//  spritebatch-sapp.c
+//------------------------------------------------------------------------------
+
+/*
+ * All artwork used in this sample was released into the public domain.
+ * Luis Zuno (ansimuz):
+ * https://opengameart.org/content/gothicvania-cemetery-pack
+ * https://opengameart.org/content/gothicvania-patreons-collection
+ * Michele Bucelli (Buch):
+ * https://opengameart.org/content/shiny-hud
+ */
+
+#define SOKOL_SPRITEBATCH_IMPL
+#define SOKOL_COLOR_IMPL
+#define STB_IMAGE_IMPLEMENTATION
+#define GAMEPLAY_WIDTH (480)
+#define GAMEPLAY_HEIGHT (270)
+
+#include "sokol_time.h"
+#include "sokol_app.h"
+#include "sokol_gfx.h"
+#include "sokol_spritebatch.h"
+#include "sokol_color.h"
+#include "sokol_glue.h"
+#include "dbgui/dbgui.h"
+#include "stb/stb_image.h"
+#include "spritebatch-sapp-crt.glsl.h"
+
+typedef struct bg_tile {
+    sbatch_float2 position;
+    sbatch_rect source;
+} bg_tile;
+
+typedef struct scroll_layer {
+    sbatch_rect source;
+    sbatch_float2 position;
+    int repeat_count;
+} scroll_layer;
+
+typedef struct animation {
+    sbatch_rect source;
+    sbatch_float2 position;
+    float frame_time;
+    int frame_count;
+    int frame_index;
+    sg_color color;
+    uint32_t mask;
+} animation;
+
+static struct {
+    uint64_t last_time;
+    float accumulator;
+
+    int screen_width;
+    int screen_height;
+    int viewport_width;
+    int viewport_height;
+    int viewport_x;
+    int viewport_y;
+
+    sg_image atlas;
+    scroll_layer mountains;
+    scroll_layer graveyard;
+    scroll_layer objects;
+    scroll_layer tiles;
+
+    animation nightmare;
+    animation demon;
+
+    sg_pass_action pass_action;
+
+    sbatch_pipeline gameplay_pipeline;
+    sg_image gameplay_render_target;
+    sbatch_context gameplay_context;
+    sg_pass gameplay_pass;
+
+    sg_shader crt_shader;
+    sbatch_pipeline crt_pipeline;
+    sbatch_context crt_context;
+    crt_params_t crt_params;
+} state;
+
+static sg_image load_sprite(const char* filepath, const char* label, sg_filter filter, sg_wrap wrap) {
+    int w, h, num_channels;
+    const int desired_channels = 4;
+
+    stbi_uc* pixels = stbi_load(filepath, &w, &h, &num_channels, desired_channels);
+    assert(pixels != NULL);
+
+    sbatch_premultiply_alpha_rgba8(pixels, w * h);
+
+    const sg_image image = sg_make_image(&(sg_image_desc) {
+        .label = label,
+        .width = w,
+        .height = h,
+        .pixel_format = SG_PIXELFORMAT_RGBA8,
+        .min_filter = filter,
+        .mag_filter = filter,
+        .wrap_u = wrap,
+        .wrap_v = wrap,
+        .data.subimage[0][0] = {
+            .ptr = pixels,
+            .size = (size_t)w * h * 4
```

Comment:
This results in a sign conversion warning (only 'w' will be cast to size_t), better:

```
.size = (size_t)(w * h * 4)
```

Reply:
Done